bip-0339.mediawiki

```diff
@@ -41,7 +41,7 @@ announcing and fetching transactions.
 
 # A new WTXIDRELAY message is added, which is defined as an empty message where pchCommand == "wtxidrelay".
 # The protocol version of nodes implementing this BIP must be set to 70016 or higher.
-# The WTXIDRELAY message must be sent in response to a VERSION message from a peer whose protocol version is >= 70016, and prior to sending a VERACK.
+# The WTXIDRELAY message may be sent in response to a VERSION message from a peer whose protocol version is >= 70016, and prior to sending a VERACK. A WTXIDRELAY message received after a VERACK message MUST be ignored or treated as invalid.
```

Comment:
Is there a reason to require all >= 70016 to send a WTXIDRELAY msg even if they do not need or want it?

Reply:
Of course not.  Implementing this BIP is not a requirement for anyone!

Reply:
That's why it's a may rather than a must. The peer may opt in to wtxid relay by sending the wtxidrelay message.

Reply:
@sdaftuar @jnewbery I think perhaps you're interpreting the language differently. Here is my guess:

Suhas means "To advertize itself as having implemented this BIP, a peer must send wtxidrelay", and that a "may" here would mean that someone can implement BIP339 without sending it.

John is interpreting the "must" as unconditional, and wants to emphasize that it is allowed for someone to not opt into BIP339.

Reply:
@jnewbery I think you're confusing compatibility with implementation.

Whether software that doesn't implement this BIP remains compatible with software that does is a separate issue from defining what implementation of this BIP means.

I think the most useful way to define what it means to implement this BIP is to require that implementing software actually exchange wtxids when announcing and fetching transactions.  The alternative is absurd: why bother specifying an implementation when all software on the network was already compliant before we did anything?

I think the concern would be if it weren't clear that software could choose to not implement the BIP and still remain compatible.  But that's why there's a backwards compatibility section...  If the text there is not enough, I'd prefer we add additional text to make it clear that implementing this BIP is not inherently incompatible with txid-based relay.

Reply:
I don't think I'm confused. As @sipa points out, we're simply interpreting the language differently.

The sentence "The WTXIDRELAY message must be sent in response to a VERSION message from a peer whose protocol version is >= 70016" implies strongly to me that if a peer sends us a protocol version >= 70016, then we *must* opt in to wtxid relay. That's fine for now, when 70016 is the highest p2p version (and you can therefore opt out by using version 70015), but if there's a later version then it's restrictive - an implementation may want to opt in to whatever feature is introduced by 70017, but not opt in to wtxid relay.

Perhaps we can reword it in a way that avoids any ambiguity:

> If a node wishes to opt in to wtxid relay with a peer whose version is >= 70016, it MUST send a WTXIDRELAY message after receiving the VERSION message and before sending the VERACK message. A WTXIDRELAY message received after a VERACK message MUST be ignored or treated as invalid.

Reply:
@jnewbery Thinking more about it, I think I agree with Suhas that that interpretation makes more sense. Every BIP supports not being implemented by some software, as after all, everything is opt-in. If a BIP poses issues with other software that doesn't implement it, that's something to go in a compatibility section.

So I don't agree that the BIP (as is) should be interpreted as version 70017, in any future version, implying that WTXIDRELAY must be sent. It is only true for software that chooses to implement BIP339, and anything is free not to do that.

If we want to make that more clear, the sentence could be prefixed with "To implement this BIP, ..." - making it clear that that is the choice. Just saying "wishes to opt in to" may still be ambiguous, because then the question may arise: what does it mean to implement the BIP, but not opt in?

Reply:
ok, reverted to MUST